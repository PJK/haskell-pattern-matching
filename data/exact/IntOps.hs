module IntOps where

-- Not exhaustive and the first clause is redundant because it's always false
f :: Int -> Int
f x | 2 `div` 1 == 1 = 1

-- Second is redundant
g :: Int -> Int
g x | 2 `rem` 3 == 2 = 1
    | otherwise = 2

-- Same but with infix
h :: Int -> Int
h x | 2 `rem` 3 == 2 = 1
    | otherwise = 2
