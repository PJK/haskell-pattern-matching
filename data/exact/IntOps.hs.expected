{
    "tag": "AnalysisSuccess",
    "contents": [
        [
            [
                "f",
                {
                    "tag": "NonExhaustive",
                    "contents": [
                        [
                            [
                                {
                                    "tag": "WildcardPattern",
                                    "contents": []
                                }
                            ],
                            "Constraints:\n  ~c == False\n  ~c == 2 `div` 1 == 1\n\nSatisfiable. Model:\n  ~c = False :: Bool\n"
                        ]
                    ]
                }
            ],
            [
                "f",
                {
                    "tag": "Redundant",
                    "contents": [
                        {
                            "tag": "WildcardPattern",
                            "contents": []
                        },
                        {
                            "tag": "GuardPattern",
                            "contents": [
                                {
                                    "tag": "ConstructorPattern",
                                    "contents": [
                                        "True",
                                        []
                                    ]
                                },
                                {
                                    "tag": "BExp",
                                    "contents": {
                                        "tag": "IntBoolOp",
                                        "contents": [
                                            "IntEQ",
                                            {
                                                "tag": "IntOp",
                                                "contents": [
                                                    "IntDiv",
                                                    {
                                                        "tag": "IntLit",
                                                        "contents": 2
                                                    },
                                                    {
                                                        "tag": "IntLit",
                                                        "contents": 1
                                                    }
                                                ]
                                            },
                                            {
                                                "tag": "IntLit",
                                                "contents": 1
                                            }
                                        ]
                                    }
                                }
                            ]
                        }
                    ]
                }
            ],
            [
                "g",
                {
                    "tag": "Redundant",
                    "contents": [
                        {
                            "tag": "WildcardPattern",
                            "contents": []
                        },
                        {
                            "tag": "GuardPattern",
                            "contents": [
                                {
                                    "tag": "ConstructorPattern",
                                    "contents": [
                                        "True",
                                        []
                                    ]
                                },
                                {
                                    "tag": "BExp",
                                    "contents": {
                                        "tag": "Otherwise",
                                        "contents": []
                                    }
                                }
                            ]
                        }
                    ]
                }
            ],
            [
                "h",
                {
                    "tag": "Redundant",
                    "contents": [
                        {
                            "tag": "WildcardPattern",
                            "contents": []
                        },
                        {
                            "tag": "GuardPattern",
                            "contents": [
                                {
                                    "tag": "ConstructorPattern",
                                    "contents": [
                                        "True",
                                        []
                                    ]
                                },
                                {
                                    "tag": "BExp",
                                    "contents": {
                                        "tag": "Otherwise",
                                        "contents": []
                                    }
                                }
                            ]
                        }
                    ]
                }
            ]
        ],
        [
            [
                "f",
                [
                    [
                        [
                            {
                                "tag": "VariablePattern",
                                "contents": "~a"
                            }
                        ],
                        [
                            {
                                "tag": "WildcardPattern",
                                "contents": []
                            }
                        ]
                    ]
                ]
            ],
            [
                "g",
                [
                    [
                        [
                            {
                                "tag": "VariablePattern",
                                "contents": "~a"
                            }
                        ],
                        [
                            {
                                "tag": "WildcardPattern",
                                "contents": []
                            }
                        ]
                    ],
                    [
                        [
                            {
                                "tag": "VariablePattern",
                                "contents": "~a"
                            }
                        ],
                        [
                            {
                                "tag": "WildcardPattern",
                                "contents": []
                            }
                        ]
                    ]
                ]
            ],
            [
                "h",
                [
                    [
                        [
                            {
                                "tag": "VariablePattern",
                                "contents": "~a"
                            }
                        ],
                        [
                            {
                                "tag": "WildcardPattern",
                                "contents": []
                            }
                        ]
                    ],
                    [
                        [
                            {
                                "tag": "VariablePattern",
                                "contents": "~a"
                            }
                        ],
                        [
                            {
                                "tag": "WildcardPattern",
                                "contents": []
                            }
                        ]
                    ]
                ]
            ]
        ]
    ]
}
